manticore/core/workspace.py

```diff
@@ -0,0 +1,458 @@
+        return FilesystemStore(uri)
+    elif type == 'redis':
+        return RedisStore(uri)
+    else:
+        raise NotImplementedError("Storage type '%s' not supported.", type)
+
+class Workspace(object):
+    '''
+    A workspace maintains a list of states to run and assigns them IDs.
+    '''
+
+    def __init__(self, lock, desc=None):
+        self._store = _create_store(desc)
+        self._serializer = PickleSerializer()
+        self._last_id = 0
+        self._lock = lock
+        self._prefix = 'state_'
+        self._suffix = '.pkl'
+
+    def try_loading_workspace(self):
+        state_names = self._store.ls('{}*'.format(self._prefix))
+        def get_state_id(name):
+            return int(name[len(self._prefix):-len(self._suffix)], 16)
+        state_ids = map(get_state_id, state_names)
+
+        if not state_ids:
+            return []
+
+        self._last_id = max(state_ids) + 1
+
+        return state_ids
+
+    def _get_id(self):
+        '''
+        Get a unique state id.
+
+        :rtype: int
+        '''
+        try:
+            self._lock.acquire()
+            id = self._last_id
+            self._last_id = id + 1
+            return id
+        finally:
+            self._lock.release()
+
+    def load_state(self, state_id):
+        '''
+        Load a state from storage identified by `state_id`.
+
+        :param state_id: The state reference of what to load
+        :return: The deserialized state
+        :rtype: State
+        '''
+        return self._store.load_state(state_id)
+
+    def save_state(self, state):
+        '''
+        Save a state to storage, return identifier.
+
+        :param state_id: The state reference of what to load
+        :param bool final: Whether the state is finalized (i.e. testcase)
+        :return: New state id
+        :rtype: int
+        '''
+        id = self._get_id()
+        self._store.save_state(state, '{}{:08x}{}'.format(self._prefix, id, self._suffix))
+        return id
+
+
+
+class ManticoreOutput(object):
+    '''
+    Functionality related to producing output. Responsible for generating state summaries,
+    coverage information, etc.
+    '''
+    def __init__(self, desc=None):
+        '''
+        Create an object capable of producing Manticore output.
+
+        :param desc: A descriptor ('type:uri') of where to write output.
+        '''
+        self._store = _create_store(desc)
+        self._last_id = 0
+
+    @property
+    def uri(self):
+        return self._store.uri
+
+    def save_testcase(self, state, message=''):
+        '''
+        Save the environment from `state` to storage. Return a state id
+        describing it, which should be an int or a string.
+
+        :param State state: The state to serialize
+        :param int testcase_id: Identifier for the state
+        :param str state: Optional message to include
+        :return: A state id representing the saved state
+        '''
+
+        self.save_summary(state, message)
```

Comment:
All this should be plug-able from manticore to enable platforms that do not have FDs or configurations that didn't want to spend time maintaining a trace.